pact/policy-manager/policy-manager.pact

```diff
@@ -0,0 +1,339 @@
+  )))
+
+  (defun enforce-transfer:[bool]
+    ( token:object{token-info}
+      sender:string
+      guard:guard
+      receiver:string
+      amount:decimal )
+    (enforce-ledger)
+    (with-capability (POLICY_MANAGER)
+      (let ((policies:[module{kip.token-policy-v2}]  (at 'policies token)))
+        (map-transfer token sender guard receiver amount policies))))
+
+
+;; Sale/Escrow Functions
+  (defun enforce-sale-pact:bool (sale:string)
+    "Enforces that SALE is id for currently executing pact"
+    (enforce (= sale (pact-id)) "Invalid pact/sale id")
+  )
+
+  (defun map-escrowed-buy:bool
+    ( sale-id:string
+      token:object{token-info}
+      seller:string
+      buyer:string
+      buyer-guard:guard
+      amount:decimal
+      policies:[module{kip.token-policy-v2}]
+    )
+    (let* (
+           (escrow-account:object{fungible-account} (get-escrow-account sale-id))
+           (quote:object{quote-schema} (get-quote-info sale-id))
+           (spec:object{quote-spec} (at 'spec quote))
+           (fungible:module{fungible-v2} (at 'fungible spec))
+           (buyer-fungible-account-name:string (read-msg BUYER-FUNGIBLE-ACCOUNT-MSG-KEY))
+           (seller-fungible-account:object{fungible-account} (at 'seller-fungible-account spec))
+           (price:decimal (at 'price spec))
+           (sale-price:decimal (floor (* price amount) (fungible::precision)))
+      )
+       ;; transfer fungible to escrow account
+       (fungible::transfer-create buyer-fungible-account-name (at 'account escrow-account) (at 'guard escrow-account) sale-price)
+
+       (with-capability (ESCROW sale-id)
+         ;; Run policies::enforce-buy
+         (map-buy token seller buyer buyer-guard amount sale-id policies)
+         ;; Transfer Escrow account to seller
+         (let (
+               (balance:decimal (fungible::get-balance (at 'account escrow-account)))
+             )
+             (install-capability (fungible::TRANSFER (at 'account escrow-account) (at 'account seller-fungible-account) balance))
+             (fungible::transfer (at 'account escrow-account) (at 'account seller-fungible-account) balance)
+         )
+       )
+       true
+    )
+  )
+
+  ;;utility functions
+
+  (defun exists-quote:bool (sale-id:string)
+    @doc "Looks up quote table for quote"
+    (= (take 6 (typeof (try false (get-quote-info sale-id)))) "object")
+  )
+
+  (defun exists-msg-decimal:bool (msg:string)
+    @doc "Checks env-data field and see if the msg is a decimal"
+    (= (typeof  (try false (read-decimal msg))) "decimal")
+  )
+
+  (defun exists-msg-object:bool (msg:string)
+    @doc "Checks env-data field and see if the msg is a object"
+    (= (take 6 (typeof  (try false  (read-msg msg)))) "object")
+  )
+
+ (defun token-init (token:object{token-info} policy:module{kip.token-policy-v2})
+  (policy::enforce-init token))
+
+ (defun map-init (token:object{token-info} policy-list:[module{kip.token-policy-v2}])
+  (map (token-init token) policy-list))
```

Comment:
Not necessary but consider using a lambda instead of having to have `token-init` toplevel function:
```suggestion
  (map (lambda (p:module{kip.token-policy-v2}) (p::enforce-init token)) policy-list)
```

Reply:
If you like this apply to other `map-XXX` functions. You can also let bind the lambda if you like that style better:


```pact
(let ((lam (lambda (p:module{kip.token-policy-v2}) 
             (p::enforce-XXX foo bar)))
  (map lam policy-list))
```